File: core/src/main/scala/com/graphbrain/eco/Token.scala
package com.graphbrain.eco

import com.graphbrain.eco.TokenType.TokenType

class Token(val text: String, val ttype: TokenType) {
  override def toString = text + " <" + ttype.toString + ">"

  override def equals(obj: Any) =
    obj.asInstanceOf[Token].text == text &&
    obj.asInstanceOf[Token].ttype == ttype

  def precedence = ttype match {
    case TokenType.LPar => 3
    case TokenType.Mul => 2
    case TokenType.Div => 2
    case TokenType.Plus => 1
    case TokenType.Minus => 1
    case _ => 0
  }
}

object Token {
  def apply(test: String, ttype: TokenType) = new Token(test, ttype)
}

File: core/src/main/java/com/graphbrain/eco/Token.java
package com.graphbrain.eco;

public class Token {

    private String text;
    private TokenType ttype;

    public Token(String text, TokenType ttype) {
        this.text = text;
        this.ttype = ttype;
    }

    @Override
    public String toString() {
        return text + " <" + ttype.toString() + ">";
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Token) {
            Token t = (Token)obj;
            return t.text.equals(text) && t.ttype.equals(ttype);
        }

        return false;
    }

    public String getText() {
        return text;
    }

    public TokenType getTtype() {
        return ttype;
    }
}